pkg/gitprovider/bitbucketserver.go

```diff
@@ -0,0 +1,402 @@
+	if staticContext.PrNumber == nil {
+		return staticContext, nil
+	}
+
+	repo := *staticContext
+
+	client, err := g.getApiClient()
+	if err != nil {
+		return nil, err
+	}
+
+	pr, err := client.DefaultApi.GetPullRequest(staticContext.Id, staticContext.Name, int(*staticContext.PrNumber))
+	if err != nil {
+		return nil, err
+	}
+
+	prInfo, err := bitbucketv1.GetPullRequestResponse(pr)
+	if err != nil {
+		return nil, err
+	}
+
+	if prInfo.FromRef.Repository.Owner != nil {
+		ownerName := prInfo.FromRef.Repository.Owner.DisplayName
+		repo.Owner = ownerName
+		repo.Id = ownerName
+	}
+	repo.Name = prInfo.FromRef.Repository.Slug
+	repo.Branch = &prInfo.FromRef.DisplayID
+
+	return &repo, nil
+}
+
+func (g *BitbucketServerGitProvider) parseStaticGitContext(repoUrl string) (*StaticGitContext, error) {
+	var staticContext StaticGitContext
+
+	// optional string - '/rest/api/'
+	re := regexp.MustCompile(`(https?://[^/]+)(?:/rest/api/[^/]+)?/projects/([^/]+)/repos/([^/]+)(?:/([^/?]+))?(?:/([^/?]+))?(?:\?at=refs%2Fheads%2F([^/?]+))?`)
+	matches := re.FindStringSubmatch(repoUrl)
+
+	if len(matches) < 4 {
+		return nil, fmt.Errorf("Could not extract project key and repo name from URL: %s", repoUrl)
+	}
+
+	baseUrl := matches[1]
+	projectKey := matches[2]
+	repoName := matches[3]
+	action := matches[4]
+	// action is either 'pull-requests', 'browse', 'commits'
+	// identifier is either pull request number or path or commit SHA
+	identifier := matches[5]
+	branchName := matches[6]
+
+	staticContext.Id = projectKey
+	staticContext.Name = repoName
+	staticContext.Owner = projectKey
+	// For '.git' or repo clone over https format, refer https://community.atlassian.com/t5/Bitbucket-questions/Project-key-in-repositories-URL/qaq-p/578207
+	// and https://community.atlassian.com/t5/Bitbucket-questions/remote-url-in-Bitbucket-server-what-does-scm-represent-is-it/qaq-p/2060987
+	staticContext.Url = fmt.Sprintf("%s/scm/%s/%s.git", baseUrl, projectKey, repoName)
+	staticContext.Source = strings.TrimPrefix(baseUrl, "https://")
+
+	switch action {
+	case "pull-requests":
+		if prNumber, err := strconv.Atoi(identifier); err == nil {
+			prUint := uint32(prNumber)
+			staticContext.PrNumber = &prUint
+		}
+	case "browse":
+		if branchName != "" {
+			staticContext.Branch = &branchName
+		} else if strings.Contains(repoUrl, "browse/") {
+			if identifier != "" {
+				staticContext.Path = &identifier
+			}
+		} else if strings.Contains(repoUrl, "browse?") {
+			if strings.Contains(repoUrl, "at=refs%2Fheads%2F") {
+				parts := strings.Split(repoUrl, "at=refs%2Fheads%2F")
+				if len(parts) == 2 {
+					branchName = parts[1]
+					staticContext.Branch = &branchName
+				}
+			}
+		}
+	case "commits":
+		if identifier != "" {
+			staticContext.Sha = &identifier
+		} else if strings.Contains(repoUrl, "commits?until=") {
+			parts := strings.Split(repoUrl, "commits?until=")
+			if len(parts) == 2 {
+				sha := parts[1]
+				staticContext.Sha = &sha
```

Comment:
You need to set `staticContext.Branch` to the SHA as well so Daytona knows it needs to checkout the commit.

When I created a repo from a commit URL, it got cloned to the main branch instead of the specified commit.
![Screenshot 2024-06-26 at 09 37 13](https://github.com/daytonaio/daytona/assets/26512078/5bb6864a-023d-48d2-bde8-5f1a20a1cf31)

Reply:
added the mentioned fix, not able to test it, got some issues in serve command after rebasing with main branch (all testing i did was with 2nd last commit changes, rebase was done after it). 
Existing test case had to be updated for this change.

Reply:
@abhishek818 you probably didn't comply with the breaking changes. The easiest method to go around it so to remove the entire config with `rm -rf ~/.config/daytona`.

![Screenshot 2024-06-26 at 15 34 43](https://github.com/daytonaio/daytona/assets/26512078/7e7a6b24-3e94-4634-aa1a-159e479676b7)

The branch again didn't go through it seems. I created the project from https://bitbucket.daytona.io/projects/DAYT/repos/test/commits/8c4472d7e949f6211ff6cba7fab762a2b1a3aa5f\#test.txt

I see the issue, your test case here checks for `Branch: nil` instead of `Branch: COMMIT_SHA`
![Screenshot 2024-06-26 at 15 35 36](https://github.com/daytonaio/daytona/assets/26512078/2f277952-6268-4de4-9bdb-b45db8caad0d)

Reply:
oh, fixed now, I forgot there were 2 types of commit url : one with 'until=' and other with 'commits/commitId'

Reply:
Seems to work now. Nice job